addons/rating/models/mail_thread.py

```diff
@@ -23,7 +23,45 @@ def unlink(self):
         return result
 
     def _get_message_create_valid_field_names(self):
-        return super()._get_message_create_valid_field_names() | {'rating_ids'}
+        return super()._get_message_create_valid_field_names() | {"rating_id", "rating_value"}
+
+    def _message_create(self, values_list):
+        """Override of _message_create rather than message_post to be just after create
+        but before any post-processing."""
+        rating_id_by_message_index = (values.pop("rating_id", None) for values in values_list)
+        rating_value_by_message_index = (values.pop("rating_value", None) for values in values_list)
+        rating_ids = (rating_id for rating_id in rating_id_by_message_index if rating_id)
+        messages = super()._message_create(values_list)
+        for message, rating, rating_value in zip(
+            messages,
+            (
+                self.env["rating.rating"].browse(rating_id).with_prefetch(rating_ids)
+                for rating_id in rating_id_by_message_index
+            ),
+            rating_value_by_message_index,
+        ):
+            if (
+                rating
+                and rating.partner_id
+                and rating.partner_id == message.author_id
+                and rating.res_model == message.model
+                and rating.res_id == message.res_id
+            ):
+                # sudo: rating.rating - current user can link their rating to their message
+                rating.sudo().message_id = message.id
+            if rating_value:
+                rating_vals = {
+                    "consumed": True,
+                    "feedback": tools.html2plaintext(message.body),
+                    "message_id": message.id,
+                    "partner_id": self.env.user.partner_id.id,
+                    "rating": float(rating_value) if rating_value is not None else False,
+                    "res_id": self.id,
+                    "res_model_id": self.env["ir.model"]._get_id(self._name),
+                }
+                # sudo: rating.rating - current user can create a rating when posting a message
+                self.env["rating.rating"].sudo().create(rating_vals)
```

Comment:
Note: this effectively means (even before these PR) that any new message can be created together with a rating. It seems like we would like more control over that, probably at the model level.

The options I see should be: free create (rare? or non-existent?), one per user per record (most of the time: product, slides, etc), only at the end of flows (livechat, task,etc).